doc/papers/ocaml-2021/memo.md

```diff
@@ -107,19 +105,22 @@ approach used by Tenacious is conceptually simple but introduces a delay between
 the creation of a dependency cycle and its detection.
 
 Compared to the incremental computation libraries mentioned above, the current
-implementation of Memo makes one controversial design choice. Incremental,
-Adapton and Tenacious are all *push based*, i.e., they trigger recomputation
-starting from the leaves of the computation graph. Memo is *pull based*:
-`invalidate` calls merely mark leaves (cells) as out of date, but recomputation
-is driven from the top-level `run` calls. The main drawback of this approach is
-that Memo needs to traverse the whole graph on each rebuild. The main benefits
-are: (i) Memo doesn't need to store reverse dependencies, which saves space and
-eliminates various garbage collection pitfalls; (ii) the pull based approach
-makes it easier to avoid "spurious" recomputations, where a node is recomputed
-but subsequently becomes unreachable from the top due to a new dependency
-structure. We may reconsider this design choice in future, but for now
-traversing the whole graph on each rebuild isn't a bottleneck even for
-large-scale builds where graphs contain millions of nodes.
+implementation of Memo makes one unusual design choice. Incremental, Adapton and
+Tenacious are all *push based*, i.e., they trigger recomputation starting from
+the leaves of the computation graph. Memo is *pull based*: `invalidate` calls
+merely mark leaves (cells) as out of date, but recomputation is driven from the
+top-level `run` calls. The main drawback of our approach is that Memo needs to
+traverse the whole graph on each rebuild. The main benefits are: (i) Memo
```

Comment:
It would be nice to explain why this is necessary.

Reply:
You mean, explain why traversing the whole graph is necessary?

Yes, indeed. Although, perhaps necessary is a too strong word. I have some ideas I'd like to experiment with where we avoid traversing some parts of the graph, at least in the common case where the user keeps editing the very same file over and over again -- in this case Memo can speculatively guide the traversal towards the same part of the tree.